tests/test_query.py

```diff
@@ -451,11 +456,93 @@ def test_should_use_nested_prefetch_related_while_also_selecting_only_required_f
     optimized_items = qs.prefetch_related(
         Prefetch(
             'children',
-            queryset=Item.objects.only('id'),
+            queryset=Item.objects.only('id', 'parent_id').prefetch_related(
+                Prefetch(
+                    'children',
+                    queryset=Item.objects.only('id', 'parent_id'),
+                )
+            ),
+        ),
+    )
+
+@pytest.mark.django_db
+def test_should_check_reverse_relations_add_foreign_key():
+    info = create_resolve_info(schema, '''
+        query {
+            items {
+                otmItems {
+                    id
+                }
+            }
+        }
+    ''')
+    qs = Item.objects.all()
+    items = gql_optimizer.query(qs, info)
+    optimized_items = qs.prefetch_related(
+        Prefetch(
+            'otm_items',
+            queryset=RelatedOneToManyItem.objects.only('id', 'item_id'),
+        ),
+    )
+    assert_query_equality(items, optimized_items)
+
+    for i in range(10):
+        the_item = Item.objects.create(name='foo')
+        for k in range(10):
+            related_item = RelatedOneToManyItem.objects.create(name='bar{}{}'.format(i,k), item=the_item)
+
+    with CaptureQueriesContext(connection) as expected_query_capture:
+        for i in items:
+            for k in i.otm_items.all():
+                pass
+
+    with CaptureQueriesContext(connection) as optimized_query_capture:
+        for i in optimized_items:
+            for k in i.otm_items.all():
+                pass
+
+    assert len(optimized_query_capture.captured_queries) == 2
+    assert len(expected_query_capture) == len(optimized_query_capture)
+
+@pytest.mark.django_db
+def test_should_check_reverse_relations_add_foreign_key():
+    info = create_resolve_info(schema, '''
+        query {
+            items {
+                otmItems {
+                    id
+                }
+            }
+        }
+    ''')
+    qs = Item.objects.all()
+    items = gql_optimizer.query(qs, info)
+    optimized_items = qs.prefetch_related(
+        Prefetch(
+            'otm_items',
+            queryset=RelatedOneToManyItem.objects.only('id', 'item_id'),
         ),
     )
     assert_query_equality(items, optimized_items)
 
+    for i in range(10):
```

Comment:
This piece of code is not clear at first view. Now I realize what it does. Maybe put a comment about what's happening or say something like "When the query is not optimized, there will be an additional 100 queries because each Item will need 10 queries for each RelatedOneToManyItem in order to fetch the needed data".

I think it's not necessary to create ten items. With one it would be enough for test correctness.

Reply:
Added a comment in b989cb4aff499fb778054787d8539c0452e7f61b